### PR TITLE
Coverup/Blanking for sensitive subjects data displayed in the management dashboard

### PR DESCRIPTION
**Pull request by sindelio**
Tested in Postman and via our test suite. 
There was no need to make a specific test for the covering up of the data because it is already part of the expectation in other tests. Although we could make one for clarity, what do you guys think?

### CODE REVIEW

File: cg/src/domains/management/subjects/subjects.controller.js [hunk 1]
```diff
@@ -7,7 +18,16 @@ class SubjectsController {
 
   async listSubjects(req, res) {
     const subjectsPage = await this.subjectsService.listSubjects(req.query.page);
-    res.json(subjectsPage);
+    if (subjectsPage.hasOwnProperty('error')) {
```
Comment (sindelio) [inline cg/src/domains/management/subjects/subjects.controller.js]: I need to fix this line, although it's working, a subject can insert an 'error' key in his personal data and that would make his data not be blanked when displayed in the management dashboard. I can think of 2 fixes: 
- Reserve the error key to the system, not allowing a subject's personal data to have an  error key
- Do a more strict check, with the contents of the error field. this does not really solve the issue, but will make it less likely to the point that only someone trying to hack the system will be able to do it
I want to hear your opinions on this

File: cg/src/domains/management/subjects/subjects.controller.js [hunk 2]
```diff
@@ -1,4 +1,15 @@
 const SubjectsService = require('./subjects.service');
+const _ = require('underscore');
+
+function coverString(string) {
+  var firstCharacter = string[0];
```
Comment (Jakedavies) [inline cg/src/domains/management/subjects/subjects.controller.js]: use let

Comment (Jakedavies) [reply cg/src/domains/management/subjects/subjects.controller.js]: actually all these variables should be const

File: cg/src/domains/management/subjects/subjects.controller.js [hunk 3]
```diff
@@ -1,4 +1,15 @@
...
+  var lastCharacter = string[string.length - 1];
+  var coveredString = new Array(string.length).fill('*');
+  coveredString[0] = firstCharacter;
+  coveredString[string.length - 1] = lastCharacter;
+  coveredString = coveredString.join('');
```
Comment (Jakedavies) [inline cg/src/domains/management/subjects/subjects.controller.js]: use a  new variable here, no reason to overwrite

File: cg/src/domains/management/subjects/subjects.controller.js (hunk 1 above)
Comment (Jakedavies) [reply cg/src/domains/management/subjects/subjects.controller.js]: Why are we checking the output of the service layer for an error object at all? I would expect that any error case at the service layer is going to throw an error and then bubble out to the top level?

Comment (sindelio) [reply cg/src/domains/management/subjects/subjects.controller.js]: This shows that I don't understand well how we handle errors in the project yet. I will try that out

File: cg/src/domains/management/subjects/subjects.controller.js [hunk 4]
```diff
@@ -1,4 +1,15 @@
 const SubjectsService = require('./subjects.service');
+const _ = require('underscore');
+
+function coverString(string) {
+  const firstCharacter = string[0];
+  const lastCharacter = string[string.length - 1];
+  let coveredString = new Array(string.length).fill('*');
```
Comment (Jakedavies) [inline cg/src/domains/management/subjects/subjects.controller.js]: This is mostly a small note but I would like you to fix it so you understand. 

You can make this `coveredString` variable a const, array items can be reassigned with a const

File: cg/src/domains/management/subjects/subjects.controller.js [hunk 5]
```diff
@@ -1,4 +1,15 @@
...
+  coveredString[0] = firstCharacter;
+  coveredString[string.length - 1] = lastCharacter;
+  coveredString = coveredString.join('');
```
Comment (Jakedavies) [inline cg/src/domains/management/subjects/subjects.controller.js]: coveredString actually becomes a completed different object here (it goes from an array to a string), so we should not re-use the variable 

In the interest of code clarity and readability I would suggest you create a new variable at the stage you are getting the final string, and return that.

Comment (Jakedavies) [reply cg/src/domains/management/subjects/subjects.controller.js]: Using a new variable here also allows the variable above you declared with `let` to be a `const`

Comment (sindelio) [reply cg/src/domains/management/subjects/subjects.controller.js]: Hmm I see, it's interesting that I can change a const array after the first assignment, didn't knew that. I really want to comprehend more how Javascript works

File: cg/src/domains/management/subjects/subjects.controller.js [hunk 6]
```diff
@@ -1,4 +1,15 @@
 const SubjectsService = require('./subjects.service');
+const _ = require('underscore');
+
+function coverString(string) {
+  const firstCharacter = string[0];
+  const lastCharacter = string[string.length - 1];
+  const fullyCoveredArray = new Array(string.length).fill('*');
+  fullyCoveredArray[0] = firstCharacter;
+  fullyCoveredArray[string.length - 1] = lastCharacter;
+  let partiallyCoveredString = fullyCoveredArray.join('');
```
Comment (Jakedavies) [inline cg/src/domains/management/subjects/subjects.controller.js]: this should be const since it never gets re-assigned

**Comment by Jakedavies** [comment]
@sindelio go ahead and merge after you change that let to a const

**Comment by sindelio** [comment]
Ok

**Pull request merged**